### ISSUE TITLE
Some repository layout and build consideration to achieve the merge of Unvanquished and unvanquished_src.dpkdir repositories

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#1371]
Currently, the game code is tracked in `Unvanquished` repository, and the related assets are tracked in `unvanquished_src.dpkdir`.
`Unvanquished` is built with CMake, while `unvanquished_src.dpkdir` is built with Urcheon, merging the two build processes are not really an option, the second one, for example, has to be able to produce incremental updates and manage versions.

The fact `unvanquished_src.dpkdir` is not a folder of `Unvanquished` makes possible to make it buildable the exact same way the others `.dpkdir` are, and does not overly complicate the ability for any tools loading . dpkdir files to find them (only one pakpath is needed), while at the same time it can still live as a submodule of `Unvanquished` and seen there as a subfolder.

Outside of the various limitations and pain that comes when dealing with submodules, I discovered a new problem when building the 0.52 rc2 package: `Unvanquished` had one commit reference but `unvanquished_src.dpkdir` had another. The version string of the universal zip is expected to be the version of `Unvanquished` build, and it is expected the `unvanquished_src.dpkdir` has the same version. If I'm right the release validation tool checks for that. When doing conventionnal releases, we use tags, and use the same tags on both `Unvanquished` and `unvanquished_src.dpkdir`, so both the universal sip and the dpk shares the same version string. But when producing non-tagged builds, the version strings mismatches.

What I've done for the 0.52rc2 package is that I used `Unvanquished` version string on the `unvanquished` dpk. Because that `Unvanquished` reference was storing the `unvanquished_src.dpkdir` submodule reference, one can recover the original commit of `unvanquished_src.dpkdir` by looking at the commit on `Unvanquished`. The references are not lost, but this is far from perfect, and it requires a manual step.

Note that we currently still need to manually import the `.nexe` files into the `.dpk` anyway.

I then thought about this solution:

- move all the sources and library submodule from `Unvanquished` repository in a subfolder (like `code` or something like that),
- import all assets from `unvanquished_src.dpkdir` at the root,
- rename `Unvanquished` as `unvanquished_src.dpkdir`, drop previous `unvanquished_src.dpkdir`

Add an install step for the gamecode, installing the nexe to a subfolder at the root of that new `unvanquished_src.dpkdir`, let's say `game`.

Building the `unvanquished_src.dpkdir` with CMake will produce and install the nexe in the right folder, then building `unvanquished_src.dpkdir` with Urcheon will produce the `.dpk`, dealing with non-code assets, partial builds, versioning, packaging, etc. The `.nexe` files would just be added to the `.dpk` file like any other file.

Note: Xonotic is already doing thins that way (with other tools than CMake and Urcheon), the game code being in `xonotic-data.pk3dir`, their `xonotic` repository just being a parent repository for everything (engine, game code, assets, various tools…).

Migrating to that new layout would be a huge effort, but that would probably solve most of our issues. We do that effort at the same time we may do that other effort that was talked about previously: to extract the actual game code history from the `Unvanquished` repository (i.e. without any trace in history of the engine, like the `Dæmon` repository was extracted without trace of `Unvanquished` in all previous commits).

This may require to modify some bits in engine, like not loading the `.nexe` from the root directory of the `.dpk` archives, but from a subfolder of them.

The layout will look like somewhat like this:

```
about/              # from unvanquished_src.dpkdir
bots/               # from unvanquished_src.dpkdir
build/              # CMake workspace, ignored by git and Urcheon
code/src/           # from Unvanquished
code/libs/          # from Unvanquished
configs/            # from unvanquished_src.dpkdir
emoticons/          # from unvanquished_src.dpkdir
fonts/              # from unvanquished_src.dpkdir
game/*game.nexe     # output of cmake && make install, ignored by Git but tracked by Urcheon
gfx/                # from unvanquished_src.dpkdir
icons/              # from unvanquished_src.dpkdir
lights/             # from unvanquished_src.dpkdir
models/             # from unvanquished_src.dpkdir
presets/            # from unvanquished_src.dpkdir
scripts/            # from unvanquished_src.dpkdir (materials)
sound/              # from unvanquished_src.dpkdir
translation/        # from unvanquished_src.dpkdir
ui/                 # from unvanquished_src.dpkdir
```

**Comment by illwieckz** [comment]
One shortcoming is that I don't see any way to do something similar to `./daemon -pakpath ../pkg` like we can currently do from the `build/` directory.

**Comment by slipher** [comment]
I'm not a fan of inverting the layout to make all the stuff in unvanquished_src.dpkdir top-level. What's wrong with keeping the layout as it is now except making unvanquished_src.dkdir a normal directory instead of a submodule? In that case, instead of directing Urcheon to ignore the `code` directory (and probably a bunch of other junk e.g. CI build scripts), you'd need to teach it to use a pak base directory different from the Git repository root.

**Comment by Gireen** [comment]
Beforehand i don't know much how the build process works or what external tools have access to unvanquished .dpk files.
So my naive question is, cant we just run CMake first and then Urcheon? for packaging, building the vms is needed anyway. 
Would it make a difference if images would not move repositorys? That way the dpk.dir would need no additional preparation like image compression.

**Comment by bmorel** [comment]
> So my naive question is, cant we just run CMake first and then Urcheon?

I don't see why this would not be possible. It would even be possible to have cmake executing Urcheon.

One issue that was not said (neither in this discussion nor in related ones AFAICT), and that is very annoying currently, in the process of 0.53 release is that it makes it hard to test UI changes, and with the library update this is very, very needed. Consider the fact we're not even be using the latest RMLUI version! This means that the pain will continue.
Not only it makes things harder to test, because `0.53/sync` depends on 0.53 daemon engine, but it also makes, as was said more than once, improvements harder to do.

I honestly don't understand why having `unvanquished_src.dpkdir` as a simple sub-folder of `unvanqushed` repo breaks Urcheon.
One thing which is said in Urcheon's description is this:

> This toolset was initially developed for the Interstellar Oasis initiative.

What this means to me is, Urcheon is a tool designed to help porting maps from tremulous to unvanquished, and was not initially thought to be used for the gamelogic's "configuration".
I do not deny it's usefulness, but I fail to see why this tool must be used for gamelogic's packing, at the cost of making development much harder. Development is hard enough, removing trouble-making systems would really not be a bad thing.

I also think a solution *must* be implemented right after 0.53 is released.

One advantage that could be earned if we got Cmake (through CPack, maybe?) able to build dpk from the subdir is that people hosting their mods would be able to use a single tool for the whole process.
The same way mappers don't like to have to run another tool to complete the map (navmeshes, minimap...) developers don't really enjoy depending on another, non-standard tool to build and deploy.
And, honestly, [Urcheon](https://github.com/DaemonEngine/Urcheon) does not feels trivial to use to me, with all those stages.
I'm happy to not have to deal with all the cruft in build process like compiling files, marking archives of libs, linking all the stuff in the correct order, stripping files, and I do not want to have to deal with Urcheon's stages. I just run "cmake .. && make" to get working things, and am happy like this (I remember playing with linker scripts in the past... it's not a fun game!)

Merging the 2 repositories would also allow to implement unit testing that verifies that the gamelogic "configuration" is correct directly in the "make" process, instead of needing to launch the game (which is not really a fast thing).
Files which could more or less easily receive such tools are: `*.cfg`, `*.bt`, at the very least. It's possible that @necessarily-equal  could know how to do some testing on the GUI files, too.
This testing could be integrated either directly in cmake, as part of the default build system, and this would really help already, but it would also then be able to be integrated in CI/CD, which would imply more robust repo.
Keeping the repo split means this unit testing is impossible or almost, because the API only lies in the C++ side, and API knowledge is required to check the syntax is correct.

**Comment by slipher** [comment]
unvanquished_src.dpkdir was converted from a submodule to a directory without changing the layout.
